src/apps/phoneme-converter.man
.TH PHONEME-CONVERTER 1 "FEBRUARY 2014" "Cainteoir Text-to-Speech" "Cainteoir Manuals"
.SH NAME
phoneme-converter \- convert from one phoneme transcription to another
.SH SYNOPSIS
.P
.B phoneme-converter [OPTION..]
.I FROM
.I TO
.I TRANSCRIPTION
.P
.B phoneme-converter [OPTION..]
.I FROM
.I TO
.P
.B phoneme-converter --chart
.I PHONEMESET
.SH DESCRIPTION
.B metadata
Extract the metadata from ePub, RDF/XML, HTML and other documents,
producing RDF statements or other metadata formats.
.SH OPTIONS
.IP "-a ACCENT, --accent ACCENT"
Use the phoneme-to-phoneme rules for the specified accent to convert the phonemes
produced by the FROM phoneme set. This is the name of a supported accent (e.g.
en-GB-rp or cmu), not a path to a phoneme-to-phoneme conversion file.
.IP "-c, --chart"
Generate a phoneme chart for the specified phonemeset.
.IP "-f, --features"
Show the phonetic features along with the transcription.
.IP "-h, --help"
Show a command-line option usage help message.
.IP "-M PHONEME_MAP, --phoneme-map PHONEME_MAP"
Use the phoneme-to-phoneme rules in PHONEME_MAP to convert the phonemes
produced by the FROM phoneme set.
.IP "--no-pauses"
Do not process pause phonemes.
.IP "-s, --separate"
Display each phoneme on a new line.
.IP "--syllable-stress"
Adjust the placement of stresses such that they are on syllable boundaries.
.IP "--vowel-stress"
Adjust the placement of stresses such that they are on vowels (e.g. eSpeak,
arpabet). Although eSpeak supports stresses placed on consonants, it places them
on vowels in its dictionaries and pronunciation rule files.
.SH AUTHOR
Reece H. Dunn <msclrhd@gmail.com>


NEWS

-------------------------------------------------------------------------------
VERSION 0.8 - Friday June 1st 2012 [Jun 2012]

  *  define an archive interface with a zip file implementation
  *  report the filename as a toc entry if no toc is present
  *  move the UTF-8 helpers to a common location
  *  optimize decoding of utf-8 and us-ascii encodings
  *  do not load a new decoder when switching to the same encoding
  *  switch document parsing from an event model to a reader model

applications:
  *  doc2doc: a simple command-line tool for converting documents from
     one format to another

document formats:
  *  pdf:     support PDF documents using the Poppler library
  *  (x)html: support extracting metadata from several meta tag names
  *  zip:     support reading generic zipped document collections
  *  ePub:    make the ePub parser use the archive interface

-------------------------------------------------------------------------------
VERSION 0.7 - Friday March 30th 2012 [Mar 2012]

  *  use BCP 47 for interpreting and comparing language tags
  *  translate the language and region names using the iso-codes package
  *  support for espeak using installed mbrola voices

document formats:
  *  (x)html: report h1..h6 sections as table of content items
  *  email:   improve email content detection and support newsgroup information
  *  rdf/xml: support property attributes on an empty property element
  *  xml:     support entities in DTD blocks

bug fixes:
  *  fix detection of html that is valid xml, but is not xhtml
  *  improve detection of control characters when identifying binary data

cleanup/refactoring:
  *  simplified and cleaned up the RDF model and query APIs
  *  use std::shared_ptr consistently over std::tr1::shared_ptr
  *  use std::make_shared instead of allocating the object explicitly
  *  remove the libxml2 dependency and make all xml-based parsers use xmlreader

-------------------------------------------------------------------------------
VERSION 0.6 -- Sunday January 29th 2012 [Jan 2012]

  *  languages.rdf -- represent ISO-639-2/B, ISO-639-2/T, ISO-639-3, ISO-639-5,
     espeak and wikimedia language codes;
  *  add document format implementation status documentation.

xmlreader xml parsing support:
  *  parse entities in attribues;
  *  support single-quote and html-style attributes;
  *  parse xml namespaces;
  *  unify xml and rdf namespace/uri support;
  *  support element/attribute mapping to context entry objects.

document parsers:
  *  epub  -- support epub 3 @refines and @datatype metadata;
  *  html  -- support all known implicit close tags and recognise more elements;
  *  ocf   -- use the xmlreader api and expose a reader api;
  *  opf   -- use the xmlreader api.

bug fixes:
  *  release the file handle when finished recording ogg files.

-------------------------------------------------------------------------------
VERSION 0.5 -- Monday November 21st 2011 [Nov 2011]

  *  use the shared-mime-info database for MIME type detection.
  *  use iconv for character encoding conversions.
  *  support all Content-Transfer-Encoding types for MIME headers, including base64.
  *  improve audio error handling during reading/recording.
  *  support more language code mappings for new espeak voices.
  *  support UND as a language identifier (for Calibri eBooks).
  *  support XML encoded HTML without an associated xmlns.

-------------------------------------------------------------------------------
VERSION 0.4 -- Tuesday July 26th 2011 [Jul 2011]

file formats:
  *  support ncx-based table of content in epub documents;
  *  support epub 3.0 metadata;
  *  basic support for SSML.

cainteoir application and engine:
  *  support listing a document's table of contents and read/record a
     specified content range;
  *  support setting the voice reading speed, pitch and volume.

-------------------------------------------------------------------------------
VERSION 0.3 -- Friday July 1st 2011 [Jun 2011]

file formats:
  *  support for single-file HTML pages (MHTML, MHT);
  *  support for RTF documents;
  *  enhanced HTML support.

bug fixes:
  *  don't segfault if the OPF document is missing from an epub file.

cainteoir:
  *  improved total reading time estimating heuristics.

-------------------------------------------------------------------------------
VERSION 0.2 -- Sunday May 29th 2011 [May 2011]

file formats:
  *  support email (RFC822) mbox format;
  *  basic support for RTF documents.

metadata:
  *  add rdfs:domain, rdfs:range and OWL markup to the text-to-speech schema;
  *  install the text-to-speech RDF schema used by the Cainteoir Engine;
  *  extract language and title metadata from HTML documents;
  *  extract metadata from email From and Subject headers.

cainteoir:
  *  list supported document and audio formats in the --help output;
  *  don't output progress when writing audio data to stdout;
  *  don't crash when using the default voice (no dc:language metadata).

-------------------------------------------------------------------------------
VERSION 0.1 -- Wednesday April 27th 2011 [Apr 2011]

cainteoir:
  *  command-line text-to-speech program;
  *  support for using espeak text-to-speech voices;
  *  listen to text, html and epub documents on the command line;
  *  record text, html and epub documents to ogg/vorbis and wave audio;
  *  see the reading/recording progress and (estimated) total time;
  *  preserve metadata from document to audio (currently epub to ogg/vorbis).

metadata:
  *  extract metadata from epub, smil, rdf/xml and other documents;
  *  generate RDF N-Triples
  *  generate RDF Turtle statements
  *  generate VorbisComment metadata


src/apps/phoneme-converter.cpp
/* Convert phonemes from one transcription scheme to another.
 *
 * Copyright (C) 2013-2014 Reece H. Dunn
 *
 * This file is part of cainteoir-engine.
 *
 * cainteoir-engine is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * cainteoir-engine is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with cainteoir-engine.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "config.h"
#include "compatibility.hpp"
#include "i18n.h"
#include "options.hpp"

#include <cainteoir/phoneme.hpp>

namespace tts = cainteoir::tts;
namespace ipa = cainteoir::ipa;

enum class phoneme_mode
{
	joined,
	separate,
	broad_markers,
	narrow_markers,
	espeak_markers,
};

void print_phonemes(std::shared_ptr<tts::phoneme_reader> &aFrom,
                    std::shared_ptr<tts::phoneme_writer> &aTo,
                    const std::shared_ptr<cainteoir::buffer> &aTranscription,
                    tts::stress_type aStress)
{
	ipa::phonemes phonemes;
	aFrom->reset(aTranscription);
	while (aFrom->read())
		phonemes.push_back(*aFrom);

	tts::make_stressed(phonemes, aStress);

	aTo->reset(stdout);
	for (const auto &phoneme : phonemes)
		aTo->write(phoneme);
	aTo->flush();
}

void print_open_marker(phoneme_mode aMode)
{
	switch (aMode)
	{
	case phoneme_mode::broad_markers:
		fprintf(stdout, "/");
		break;
	case phoneme_mode::narrow_markers:
		fprintf(stdout, "[");
		break;
	case phoneme_mode::espeak_markers:
		fprintf(stdout, "[[");
		break;
	}
}

void print_close_marker(phoneme_mode aMode)
{
	switch (aMode)
	{
	case phoneme_mode::broad_markers:
		fprintf(stdout, "/");
		break;
	case phoneme_mode::narrow_markers:
		fprintf(stdout, "]");
		break;
	case phoneme_mode::espeak_markers:
		fprintf(stdout, "]]");
		break;
	}
}

void print_phonemes(std::shared_ptr<tts::phoneme_reader> &aFrom,
                    std::shared_ptr<tts::phoneme_writer> &aTo,
                    const std::shared_ptr<cainteoir::buffer> &aTranscription,
                    phoneme_mode aMode,
                    bool aNoPauses,
                    bool aShowFeatures)
{
	auto feat = tts::createPhonemeWriter("features");
	bool need_open_marker = true;

	aFrom->reset(aTranscription);
	aTo->reset(stdout);
	feat->reset(stdout);
	while (true)
	{
		try
		{
			if (!aFrom->read())
			{
				aTo->flush();
				return;
			}
		}
		catch (const tts::phoneme_error &e)
		{
			fflush(stdout);
			fprintf(stderr, "\nerror: %s\n", e.what());
			fflush(stderr);
			continue;
		}

		if (aFrom->get(ipa::phoneme_type) == ipa::pause && aNoPauses)
			aTo->flush();
		else
		{
			if (need_open_marker)
			{
				print_open_marker(aMode);
				need_open_marker = false;
			}
			if (aFrom->get(ipa::phoneme_type | ipa::length) == (ipa::pause))
			{
				aTo->flush();
				print_close_marker(aMode);
				need_open_marker = true;
			}
			aTo->write(*aFrom);
			if (aShowFeatures)
			{
				fputc('\t', stdout);
				feat->write(*aFrom);
			}
			if (aMode == phoneme_mode::separate && !aFrom->get(ipa::joined_to_next_phoneme))
				fputc('\n', stdout);
		}
	}
}

int main(int argc, char ** argv)
{
	try
	{
		tts::stress_type stress = tts::stress_type::as_transcribed;
		phoneme_mode mode = phoneme_mode::joined;
		bool no_pauses = false;
		bool show_features = false;
		const char *phoneme_map = nullptr;
		const char *accent = nullptr;

		const option_group general_options = { nullptr, {
			{ 's', "separate", bind_value(mode, phoneme_mode::separate),
			  i18n("Display each phoneme on a new line") },
			{ 'f', "features", bind_value(show_features, true),
			  i18n("Show the features along with the transcription") },
			{ 0, "no-pauses", bind_value(no_pauses, true),
			  i18n("Do not process pause phonemes") },
			{ 'M', "phoneme-map", phoneme_map, "PHONEME_MAP",
			  i18n("Use PHONEME_MAP to convert phonemes (e.g. accent conversion)") },
			{ 'a', "accent", accent, "ACCENT",
			  i18n("Use ACCENT to convert phonemes to the specified accent") },
		}};

		const option_group stress_options = { i18n("Phoneme Stress Placement:"), {
			{ 0, "vowel-stress", bind_value(stress, tts::stress_type::vowel),
			  i18n("Place the stress on vowels (e.g. espeak, arpabet)") },
			{ 0, "syllable-stress", bind_value(stress, tts::stress_type::syllable),
			  i18n("Place the stress on syllable boundaries") },
		}};

		const option_group marker_options = { i18n("Phoneme Markers:"), {
			{ 0, "broad", bind_value(mode, phoneme_mode::broad_markers),
			  i18n("Use /.../ between phonetic transcriptions") },
			{ 0, "narrow", bind_value(mode, phoneme_mode::narrow_markers),
			  i18n("Use [...] between phonetic transcriptions") },
			{ 0, "espeak", bind_value(mode, phoneme_mode::espeak_markers),
			  i18n("Use [[...]] between phonetic transcriptions") },
		}};

		const std::initializer_list<const char *> usage = {
			i18n("phoneme-converter [OPTION..] FROM TO TRANSCRIPTION"),
			i18n("phoneme-converter [OPTION..] FROM TO"),
		};

		if (!parse_command_line({ general_options, stress_options, marker_options }, usage, argc, argv))
			return 0;

		if (argc != 2 && argc != 3)
		{
			print_help({ general_options }, usage);
			return 0;
		}
		auto from = tts::createPhonemeReader(argv[0]);
		auto to   = tts::createPhonemeWriter(argv[1]);
		auto data = argc == 3 ? cainteoir::make_file_buffer(argv[2])
		                      : cainteoir::make_file_buffer(stdin);

		if (phoneme_map)
			from = tts::createPhonemeToPhonemeConverter(phoneme_map, from);
		if (accent)
			from = tts::createAccentConverter(accent, from);

		if (stress == tts::stress_type::as_transcribed)
			print_phonemes(from, to, data, mode, no_pauses, show_features);
		else
			print_phonemes(from, to, data, stress);
	}
	catch (std::runtime_error &e)
	{
		fprintf(stderr, i18n("error: %s\n"), e.what());
		return EXIT_FAILURE;
	}

	return EXIT_SUCCESS;
}
